File: dataset/CoT/CLAUDE_claude-sonnet-4-5-20250929/Task7.java

import java.io.*;
import java.net.*;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.*;
import java.util.concurrent.*;

class Task7 {
    // Server implementation
    static class ChatServer {
        private static final int PORT = 8080;
        private static final String USERS_FILE = "users.dat";
        private static final int MAX_LOGIN_ATTEMPTS = 3;
        private static final long LOCKOUT_TIME = 300000; // 5 minutes
        private static final ConcurrentHashMap<String, Integer> loginAttempts = new ConcurrentHashMap<>();
        private static final ConcurrentHashMap<String, Long> lockedAccounts = new ConcurrentHashMap<>();
        
        public static void startServer() {
            try (ServerSocket serverSocket = new ServerSocket(PORT)) {
                System.out.println("Server started on port " + PORT);
                
                while (true) {
                    Socket clientSocket = serverSocket.accept();
                    new Thread(() -> handleClient(clientSocket)).start();
                }
            } catch (IOException e) {
                System.err.println("Server error: " + e.getMessage());
            }
        }
        
        private static void handleClient(Socket socket) {
            try (
                BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream(), StandardCharsets.UTF_8));
                PrintWriter out = new PrintWriter(new OutputStreamWriter(socket.getOutputStream(), StandardCharsets.UTF_8), true)
            ) {
                String request = in.readLine();
                if (request == null || request.trim().isEmpty()) {
                    out.println("ERROR:Invalid request");
                    return;
                }
                
                String response = processRequest(request);
                out.println(response);
                
            } catch (IOException e) {
                System.err.println("Client handling error: " + e.getMessage());
            } finally {
                try {
                    socket.close();
                } catch (IOException e) {
                    System.err.println("Socket close error: " + e.getMessage());
                }
            }
        }
        
        private static String processRequest(String request) {
            // Input validation
            if (request == null || request.length() > 1000) {
                return "ERROR:Invalid request format";
            }
            
            String[] parts = request.split(":", 3);
            if (parts.length != 3 || !parts[0].equals("LOGIN")) {
                return "ERROR:Invalid request format";
            }
            
            String username = sanitizeInput(parts[1]);
            String password = parts[2];
            
            // Validate username format
            if (!isValidUsername(username)) {
                return "ERROR:Invalid username format";
            }
            
            // Check if account is locked
            if (isAccountLocked(username)) {
                return "ERROR:Account temporarily locked";
            }
            
            // Authenticate user
            if (authenticateUser(username, password)) {
                loginAttempts.remove(username);
                return "SUCCESS:Login successful";
            } else {
                incrementLoginAttempts(username);
                return "ERROR:Invalid credentials";
            }
        }
        
        private static boolean isAccountLocked(String username) {
            Long lockoutTime = lockedAccounts.get(username);
            if (lockoutTime != null) {
                if (System.currentTimeMillis() < lockoutTime) {
                    return true;
                } else {
                    lockedAccounts.remove(username);
                    loginAttempts.remove(username);
                }
            }
            return false;
        }
        
        private static void incrementLoginAttempts(String username) {
            int attempts = loginAttempts.getOrDefault(username, 0) + 1;
            loginAttempts.put(username, attempts);
            
            if (attempts >= MAX_LOGIN_ATTEMPTS) {
                lockedAccounts.put(username, System.currentTimeMillis() + LOCKOUT_TIME);
            }
        }
        
        private static boolean isValidUsername(String username) {
            if (username == null || username.isEmpty() || username.length() > 50) {
                return false;
            }
            return username.matches("^[a-zA-Z0-9_]+$");
        }
        
        private static String sanitizeInput(String input) {
            if (input == null) return "";
            return input.replaceAll("[^a-zA-Z0-9_]", "");
        }
        
        private static boolean authenticateUser(String username, String password) {
            try {
                File usersFile = new File(USERS_FILE);
                if (!usersFile.exists()) {
                    initializeUsersFile();
                }
                
                try (BufferedReader reader = new BufferedReader(new FileReader(usersFile, StandardCharsets.UTF_8))) {
                    String line;
                    while ((line = reader.readLine()) != null) {
                        String[] parts = line.split(":");
                        if (parts.length == 3) {
                            String storedUsername = parts[0];
                            String storedSalt = parts[1];
                            String storedHash = parts[2];
                            
                            if (storedUsername.equals(username)) {
                                String computedHash = hashPassword(password, storedSalt);
                                return MessageDigest.isEqual(
                                    computedHash.getBytes(StandardCharsets.UTF_8),
                                    storedHash.getBytes(StandardCharsets.UTF_8)
                                );
                            }
                        }
                    }
                }
            } catch (IOException e) {
                System.err.println("Authentication error: " + e.getMessage());
            }
            return false;
        }
        
        private static void initializeUsersFile() {
            try (PrintWriter writer = new PrintWriter(new FileWriter(USERS_FILE, StandardCharsets.UTF_8))) {
                // Create test users with hashed passwords
                String[][] users = {
                    {"user1", "password1"},
                    {"user2", "password2"},
                    {"admin", "admin123"},
                    {"testuser", "test123"},
                    {"alice", "alice456"}
                };
                
                for (String[] user : users) {
                    String salt = generateSalt();
                    String hash = hashPassword(user[1], salt);
                    writer.println(user[0] + ":" + salt + ":" + hash);
                }
            } catch (IOException e) {
                System.err.println("Error initializing users file: " + e.getMessage());
            }
        }
        
        private static String generateSalt() {
            SecureRandom random = new SecureRandom();
            byte[] salt = new byte[16];
            random.nextBytes(salt);
            return Base64.getEncoder().encodeToString(salt);
        }
        
        private static String hashPassword(String password, String salt) {
            try {
                MessageDigest md = MessageDigest.getInstance("SHA-256");
                md.update(salt.getBytes(StandardCharsets.UTF_8));
                byte[] hashedPassword = md.digest(password.getBytes(StandardCharsets.UTF_8));
                return Base64.getEncoder().encodeToString(hashedPassword);
            } catch (NoSuchAlgorithmException e) {
                throw new RuntimeException("SHA-256 algorithm not found", e);
            }
        }
    }
    
    // Client implementation
    static class ChatClient {
        private static final String SERVER_HOST = "localhost";
        private static final int SERVER_PORT = 8080;
        private static final int CONNECTION_TIMEOUT = 5000;
        
        public static String login(String username, String password) {
            // Input validation
            if (!isValidInput(username) || !isValidInput(password)) {
                return "ERROR:Invalid input format";
            }
            
            if (username.length() > 50 || password.length() > 100) {
                return "ERROR:Input too long";
            }
            
            String request = buildLoginRequest(username, password);
            return sendRequest(request);
        }
        
        private static boolean isValidInput(String input) {
            return input != null && !input.isEmpty() && !input.contains("\\n") && !input.contains("\\r");
        }
        
        private static String buildLoginRequest(String username, String password) {
            // Sanitize username but preserve password as-is for authentication
            String sanitizedUsername = username.replaceAll("[^a-zA-Z0-9_]", "");
            return "LOGIN:" + sanitizedUsername + ":" + password;
        }
        
        private static String sendRequest(String request) {
            try (Socket socket = new Socket()) {
                socket.connect(new InetSocketAddress(SERVER_HOST, SERVER_PORT), CONNECTION_TIMEOUT);
                socket.setSoTimeout(CONNECTION_TIMEOUT);
                
                try (
                    PrintWriter out = new PrintWriter(new OutputStreamWriter(socket.getOutputStream(), StandardCharsets.UTF_8), true);
                    BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream(), StandardCharsets.UTF_8))
                ) {
                    out.println(request);
                    String response = in.readLine();
                    return response != null ? response : "ERROR:No response from server";
                }
            } catch (IOException e) {
                return "ERROR:Connection failed - " + e.getMessage();
            }
        }
        
        public static void interactiveLogin() {
            try (Scanner scanner = new Scanner(System.in)) {
                System.out.print("Enter username: ");
                String username = scanner.nextLine().trim();
                
                System.out.print("Enter password: ");
                String password = scanner.nextLine();
                
                String response = login(username, password);
                System.out.println("Server response: " + response);
            }
        }
    }
    
    // Main method with test cases
    public static void main(String[] args) {
        // Start server in a separate thread
        Thread serverThread = new Thread(() -> ChatServer.startServer());
        serverThread.setDaemon(true);
        serverThread.start();
        
        // Wait for server to start
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        
        // Test cases
        System.out.println("=== Running Test Cases ===\\n");
        
        // Test case 1: Valid login
        System.out.println("Test 1 - Valid user login:");
        String response1 = ChatClient.login("user1", "password1");
        System.out.println("Response: " + response1);
        System.out.println();
        
        // Test case 2: Invalid password
        System.out.println("Test 2 - Invalid password:");
        String response2 = ChatClient.login("user1", "wrongpassword");
        System.out.println("Response: " + response2);
        System.out.println();
        
        // Test case 3: Non-existent user
        System.out.println("Test 3 - Non-existent user:");
        String response3 = ChatClient.login("nonexistent", "password");
        System.out.println("Response: " + response3);
        System.out.println();
        
        // Test case 4: Valid admin login
        System.out.println("Test 4 - Admin login:");
        String response4 = ChatClient.login("admin", "admin123");
        System.out.println("Response: " + response4);
        System.out.println();
        
        // Test case 5: Empty credentials
        System.out.println("Test 5 - Empty credentials:");
        String response5 = ChatClient.login("", "");
        System.out.println("Response: " + response5);
        System.out.println();
    }
}

File: dataset/CoT/CLAUDE_claude-sonnet-4-5-20250929/Task7.py

import socket
import hashlib
import base64
import os
import threading
import time
from typing import Dict, Tuple

class Task7:
    class ChatServer:
        PORT = 8080
        USERS_FILE = "users.dat"
        MAX_LOGIN_ATTEMPTS = 3
        LOCKOUT_TIME = 300  # 5 minutes in seconds
        
        def __init__(self):
            self.login_attempts: Dict[str, int] = {}
            self.locked_accounts: Dict[str, float] = {}
            self.lock = threading.Lock()
        
        def start_server(self):
            try:
                server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
                server_socket.bind(('localhost', self.PORT))
                server_socket.listen(5)
                print(f"Server started on port {self.PORT}")
                
                while True:
                    client_socket, address = server_socket.accept()
                    thread = threading.Thread(target=self.handle_client, args=(client_socket,))
                    thread.daemon = True
                    thread.start()
            except Exception as e:
                print(f"Server error: {e}")
        
        def handle_client(self, client_socket):
            try:
                client_socket.settimeout(5.0)
                request = client_socket.recv(1024).decode('utf-8').strip()
                
                if not request:
                    response = "ERROR:Invalid request"
                else:
                    response = self.process_request(request)
                
                client_socket.sendall((response + '\\n').encode('utf-8'))
            except Exception as e:
                print(f"Client handling error: {e}")
            finally:
                client_socket.close()
        
        def process_request(self, request: str) -> str:
            # Input validation
            if not request or len(request) > 1000:
                return "ERROR:Invalid request format"
            
            parts = request.split(':', 2)
            if len(parts) != 3 or parts[0] != 'LOGIN':
                return "ERROR:Invalid request format"
            
            username = self.sanitize_input(parts[1])
            password = parts[2]
            
            # Validate username format
            if not self.is_valid_username(username):
                return "ERROR:Invalid username format"
            
            # Check if account is locked
            if self.is_account_locked(username):
                return "ERROR:Account temporarily locked"
            
            # Authenticate user
            if self.authenticate_user(username, password):
                with self.lock:
                    self.login_attempts.pop(username, None)
                return "SUCCESS:Login successful"
            else:
                self.increment_login_attempts(username)
                return "ERROR:Invalid credentials"
        
        def is_account_locked(self, username: str) -> bool:
            with self.lock:
                lockout_time = self.locked_accounts.get(username)
                if lockout_time:
                    if time.time() < lockout_time:
                        return True
                    else:
                        self.locked_accounts.pop(username, None)
                        self.login_attempts.pop(username, None)
            return False
        
        def increment_login_attempts(self, username: str):
            with self.lock:
                attempts = self.login_attempts.get(username, 0) + 1
                self.login_attempts[username] = attempts
                
                if attempts >= self.MAX_LOGIN_ATTEMPTS:
                    self.locked_accounts[username] = time.time() + self.LOCKOUT_TIME
        
        @staticmethod
        def is_valid_username(username: str) -> bool:
            if not username or len(username) > 50:
                return False
            return all(c.isalnum() or c == '_' for c in username)
        
        @staticmethod
        def sanitize_input(input_str: str) -> str:
            if not input_str:
                return ""
            return ''.join(c for c in input_str if c.isalnum() or c == '_')
        
        def authenticate_user(self, username: str, password: str) -> bool:
            try:
                if not os.path.exists(self.USERS_FILE):
                    self.initialize_users_file()
                
                with open(self.USERS_FILE, 'r', encoding='utf-8') as f:
                    for line in f:
                        parts = line.strip().split(':')
                        if len(parts) == 3:
                            stored_username, stored_salt, stored_hash = parts
                            if stored_username == username:
                                computed_hash = self.hash_password(password, stored_salt)
                                return computed_hash == stored_hash
            except Exception as e:
                print(f"Authentication error: {e}")
            return False
        
        def initialize_users_file(self):
            try:
                users = [
                    ("user1", "password1"),
                    ("user2", "password2"),
                    ("admin", "admin123"),
                    ("testuser", "test123"),
                    ("alice", "alice456")
                ]
                
                with open(self.USERS_FILE, 'w', encoding='utf-8') as f:
                    for username, password in users:
                        salt = self.generate_salt()
                        hash_val = self.hash_password(password, salt)
                        f.write(f"{username}:{salt}:{hash_val}\\n")
            except Exception as e:
                print(f"Error initializing users file: {e}")
        
        @staticmethod
        def generate_salt() -> str:
            return base64.b64encode(os.urandom(16)).decode('utf-8')
        
        @staticmethod
        def hash_password(password: str, salt: str) -> str:
            hash_obj = hashlib.sha256()
            hash_obj.update(salt.encode('utf-8'))
            hash_obj.update(password.encode('utf-8'))
            return base64.b64encode(hash_obj.digest()).decode('utf-8')
    
    class ChatClient:
        SERVER_HOST = 'localhost'
        SERVER_PORT = 8080
        CONNECTION_TIMEOUT = 5.0
        
        @staticmethod
        def login(username: str, password: str) -> str:
            # Input validation
            if not Task7.ChatClient.is_valid_input(username) or not Task7.ChatClient.is_valid_input(password):
                return "ERROR:Invalid input format"
            
            if len(username) > 50 or len(password) > 100:
                return "ERROR:Input too long"
            
            request = Task7.ChatClient.build_login_request(username, password)
            return Task7.ChatClient.send_request(request)
        
        @staticmethod
        def is_valid_input(input_str: str) -> bool:
            return input_str is not None and input_str and '\\n' not in input_str and '\\r' not in input_str
        
        @staticmethod
        def build_login_request(username: str, password: str) -> str:
            sanitized_username = ''.join(c for c in username if c.isalnum() or c == '_')
            return f"LOGIN:{sanitized_username}:{password}"
        
        @staticmethod
        def send_request(request: str) -> str:
            try:
                client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                client_socket.settimeout(Task7.ChatClient.CONNECTION_TIMEOUT)
                client_socket.connect((Task7.ChatClient.SERVER_HOST, Task7.ChatClient.SERVER_PORT))
                
                client_socket.sendall((request + '\\n').encode('utf-8'))
                response = client_socket.recv(1024).decode('utf-8').strip()
                
                client_socket.close()
                return response if response else "ERROR:No response from server"
            except Exception as e:
                return f"ERROR:Connection failed - {e}"
        
        @staticmethod
        def interactive_login():
            username = input("Enter username: ").strip()
            password = input("Enter password: ")
            
            response = Task7.ChatClient.login(username, password)
            print(f"Server response: {response}")

def main():
    # Start server in a separate thread
    server = Task7.ChatServer()
    server_thread = threading.Thread(target=server.start_server)
    server_thread.daemon = True
    server_thread.start()
    
    # Wait for server to start
    time.sleep(1)
    
    # Test cases
    print("=== Running Test Cases ===\\n")
    
    # Test case 1: Valid login
    print("Test 1 - Valid user login:")
    response1 = Task7.ChatClient.login("user1", "password1")
    print(f"Response: {response1}\\n")
    
    # Test case 2: Invalid password
    print("Test 2 - Invalid password:")
    response2 = Task7.ChatClient.login("user1", "wrongpassword")
    print(f"Response: {response2}\\n")
    
    # Test case 3: Non-existent user
    print("Test 3 - Non-existent user:")
    response3 = Task7.ChatClient.login("nonexistent", "password")
    print(f"Response: {response3}\\n")
    
    # Test case 4: Valid admin login
    print("Test 4 - Admin login:")
    response4 = Task7.ChatClient.login("admin", "admin123")
    print(f"Response: {response4}\\n")
    
    # Test case 5: Empty credentials
    print("Test 5 - Empty credentials:")
    response5 = Task7.ChatClient.login("", "")
    print(f"Response: {response5}\\n")

if __name__ == "__main__":
    main()
